mbed-coap/sn_coap_header.h

```diff
@@ -259,6 +259,11 @@ typedef struct registration_info_ {
     uint8_t *links_ptr;             /**< Resource registration string */
 } registration_info_t;
 
+typedef struct sn_coap_blockwise_context_ {
+    uint8_t (*blockwise_payload_get)(uint32_t block_number, uint16_t block_size, /*out*/ uint16_t *payload_len, /*out*/ uint8_t **payload_ptr, void* user_context);
```

Comment:
This API needs to be specified more... Presumably payload_len should equal block_size except for last block, and that's the "end" cue? Looking at the spec, is that sufficient to set the "More" flag? Not sure you can do "block size, block size, 0".

What's the return value?

I think it's more generally conventional to pass user context as first parameter (but not if it would be inconsistent with the rest of the CoAP library).

Can block_size be assumed to be constant for an entire transfer? Or might you see it shift at the start due to negotiation? If it shifts, maybe block_offset might be better than number? uint32_t is enough to cover the maximum offset for CoAP (20 bits of number, 2^10 blocksize).

Reply:
Yes, payload_len should equal block_size for all except last block where it can be smaller. I think we need the more flag because it is possible that last payload_len is equal to block_size but no more blocks need to be sent. The return value indicates if more blocks are to come, it's the more flag basically. I need to document this.

I'm not sure if it's allowed to send full blocks and finally an empty block with more bit cleared. Need to check if spec has any limitations, gut feeling is that should be ok to do...

Server might send different size block1 during transaction and I think it should be handled around sn_coap_protocol.c:1727 (https://github.com/ARMmbed/mbed-coap/pull/38/files#diff-2561cb94630761c7e73a0f49fc743155R1727) by calculating the correct block number based on new block1 option. This would automatically work with this streaming implementation because the callback would be called with block number matching offset with new block size. 

It would probably be better to pass the actual data offset rather than block number to the callback, since most likely the callback doesn't care about block number and would anyways have to calculate the offset.

Reply:
I agree, having just data offset would likely be easier for client side. Should there be a way for client to limit the amount of data handled at time? For example, the client side may able to handle only one block of flash at time. Now API has the payload_len, but it must equal to given length.

Reply:
@TeroJaasko Do you mean should the callback be able to indicate that it can handle only smaller blocks? Ie. callback is called with block_size=1024 but let's say callback can only handle 256 size blocks it could somehow change the block_size to 256 and rest of transfer would happen using 256 size blocks?